src/riscv_machine.cpp

```diff
@@ -249,13 +249,13 @@ static uint32_t clint_read(void *opaque, uint32_t offset, int size_log2)
 #endif
 
     switch (size_log2) {
-        case DEVIO_SIZE8:
-            val = val & 0xff;
-            break;
-        case DEVIO_SIZE16:
+        case 1:
             val = val & 0xffff;
             break;
-        case DEVIO_SIZE32:
```

Comment:
(We discussed off-line, but for the record:)

Sorry, I should have caught this when you changed it previously: we don't handle sub-word reads correctly before nor after this change.  You'd need to shift and mask depending on alignment and word access.  This change deletes byte-size handling and still doesn't handle the shifts.

The fundamental problem is that the way RISCVEMU's original device models work - it's up to each device handler to deal with sub-word accesses.  Dealing with aligned sub-word reads is not too hard (but you have to remember to shift as needed).  However, dealing with sub-word writes is not simple and there's much potential for bugs.

If we truly need to support sub-word writes to the CLINT, then I think we'd need to completely change how this works.  I can prep a review and see what people thinks.

Reply:
Makes sense. I only added the paths that let me sink subword requests to get cosim working as when in cosim mode the DUT returned values overload the CLINT values modeled. I haven't tested this for standalone runs at all.